src/Ouroboros/Network/ChainFragment.hs

```diff
@@ -0,0 +1,536 @@
+foldChainFragment blk gen (ChainFragment c) = Foldable.foldl' blk gen c
+
+-- | Specification of 'foldChainFragment'.
+--
+-- Use 'foldChainFragment', as it should be faster.
+--
+-- This function is used to verify whether 'foldChainFragment' behaves as
+-- expected.
+foldChainFragmentSpec :: HasHeader block
+                      => (a -> block -> a) -> a -> ChainFragment block -> a
+foldChainFragmentSpec _blk gen Empty    = gen
+foldChainFragmentSpec  blk gen (c :> b) = blk (foldChainFragment blk gen c) b
+
+prettyPrintChainFragment :: HasHeader block
+                         => String -> (block -> String) -> ChainFragment block -> String
+prettyPrintChainFragment nl ppBlock = foldChainFragment (\s b -> s ++ nl ++ "    " ++ ppBlock b) "ChainFragment:"
+
+mapChainFragment :: (HasHeader block1, HasHeader block2)
+                 => (block1 -> block2) -> ChainFragment block1 -> ChainFragment block2
+mapChainFragment f (ChainFragment c) = ChainFragment (FT.fmap' f c)
+
+-- TODO more efficient with a fold? -> avoid constructing views
+valid :: HasHeader block => ChainFragment block -> Bool
+valid Empty = True
+valid (c :> b) = valid c && validExtension c b
+
+-- | Checks whether the first block @bSucc@ is a valid successor of the second
+-- block @b@.
+--
+-- * The 'blockPrevHash' of the @bSucc@ must match that of @b@.
+-- * The 'blockSlot' of @bSucc@ must be strictly larger than that of @b@.
+-- * The 'blockNo' of @bSucc@ must be 1 greater than that of @b@.
+--
+-- This function does not check whether any of the two blocks satisfy
+-- 'blockInvariant'.
+isValidSuccessorOf :: HasHeader block
+                   => block  -- ^ @bSucc@
+                   -> block  -- ^ @b@
+                   -> Bool
+isValidSuccessorOf bSucc b =
+    pointHash p == blockPrevHash bSucc
+ && pointSlot p <  blockSlot bSucc
+ && blockNo   b == pred (blockNo bSucc)
+  where
+    p = blockPoint b
+
+validExtension ::  HasHeader block => ChainFragment block -> block -> Bool
+validExtension c bSucc = blockInvariant bSucc && case head c of
+  Nothing -> True
+  Just b  -> bSucc `isValidSuccessorOf` b
+
+head :: HasHeader block => ChainFragment block -> Maybe block
+head (_ :> b) = Just b
+head Empty    = Nothing
+
+headPoint :: HasHeader block => ChainFragment block -> Maybe (Point block)
+headPoint (_ :> b) = Just (blockPoint b)
+headPoint Empty    = Nothing
+
+headSlot :: HasHeader block => ChainFragment block -> Maybe Slot
+headSlot c = pointSlot <$> headPoint c
+
+headHash :: HasHeader block => ChainFragment block -> Maybe (Hash block)
+headHash c = pointHash <$> headPoint c
+
+headBlockNo :: HasHeader block => ChainFragment block -> Maybe BlockNo
+headBlockNo (_ :> b) = Just (blockNo b)
+headBlockNo Empty    = Nothing
+
+
+-- | Make a list of blocks from a 'ChainFragment', in newest-to-oldest order.
+toNewestFirst :: HasHeader block => ChainFragment block -> [block]
+toNewestFirst = foldChainFragment (flip (:)) []
```

Comment:
Is this O(n log n)?

Note how the fingertree `Foldable` instance works. They do not provide a special `toList` operation, working instead with `foldMap` that goes over the structure of the tree. It reles on the default `toList` in terms of `foldMap`. So we're using 'foldChainFragment' here which uses `Foldable.foldl'`. It's worth checking what complexity this would give us and if we can do at least as well as `Foldable.toList` does for `FingerTree` which the docs claim is O(n).